### PR TITLE
ISPN-8568 ClassNotFoundException with Compat mode and custom Pojos

### PR DESCRIPTION
**Pull request by gustavonalle**
https://issues.jboss.org/browse/ISPN-8568
https://issues.jboss.org/browse/ISPN-8553

### CODE REVIEW

File: tasks/api/src/main/java/org/infinispan/tasks/SecurityActions.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      if (System.getSecurityManager() != null) {
+         return AccessController.doPrivileged(action);
+      } else {
+         return Security.doPrivileged(action);
+      }
+   }
+
+   static Configuration getCacheConfiguration(final AdvancedCache<?, ?> cache) {
+      GetCacheConfigurationAction action = new GetCacheConfigurationAction(cache);
+      return doPrivileged(action);
```
Comment (anistor) [inline tasks/api/src/main/java/org/infinispan/tasks/SecurityActions.java]: ```doPrivileged(cache::getCacheConfiguration)``` ?

File: core/src/main/java/org/infinispan/encoding/DataConversion.java [hunk 2]
```diff
@@ -180,14 +180,14 @@ private void lookupTranscoder(EncoderRegistry encoderRegistry) {
...
-   private void lookupEncoder(EncoderRegistry encoderRegistry, CompatibilityModeConfiguration compatConfig) {
+   private void lookupEncoder(EncoderRegistry encoderRegistry, CompatibilityModeConfiguration compatConfig, ClassLoader classLoader) {
       if (!compatConfig.enabled()) {
          this.encoder = encoderRegistry.getEncoder(encoderClass, encoderId);
       } else {
          if (encoderClass == null || encoderClass.equals(CompatModeEncoder.class)) {
             this.encoderClass = CompatModeEncoder.class;
             Marshaller compatMarshaller = compatConfig.marshaller();
-            this.encoder = new CompatModeEncoder(compatMarshaller);
+            this.encoder = new CompatModeEncoder(compatMarshaller,classLoader);
```
Comment (anistor) [inline core/src/main/java/org/infinispan/encoding/DataConversion.java]: WS formatting

File: tasks/api/src/main/java/org/infinispan/tasks/TaskContext.java [hunk 3]
```diff
@@ -50,7 +51,12 @@ public TaskContext marshaller(Marshaller marshaller) {
     * other caches can be obtained from the cache manager
     */
    public TaskContext cache(Cache<?, ?> cache) {
-      this.cache = Optional.of(cache);
+      Optional<Cache<?, ?>> optionalCache = Optional.of(cache);
```
Comment (anistor) [inline tasks/api/src/main/java/org/infinispan/tasks/TaskContext.java]: Optional is so so disgusting :(

Comment (gustavonalle) [reply tasks/api/src/main/java/org/infinispan/tasks/TaskContext.java]: Removing pre-existing Optional was out of scope of this PR

Comment (anistor) [reply tasks/api/src/main/java/org/infinispan/tasks/TaskContext.java]: I know... I retract my disgustment. Yuk. :)

**Comment by gustavonalle** [comment]
Test failures aren't related.

**Comment by anistor** [comment]
LGTM.

**Comment by anistor** [comment]
Integrated in master. Thanks @gustavonalle !

**Pull request closed without merging**